File: Day01/js/solution.js
const fs = require('fs');

function extractDigits(line) {
    // Mapping of digit words to their numeric values
    const digitMap = {
        zero: 0, one: 1, two: 2, three: 3, four: 4,
        five: 5, six: 6, seven: 7, eight: 8, nine: 9
    };

    // Add single digit mappings
    for (let i = 0; i <= 9; i++) {
        digitMap[i.toString()] = i;
    }

    let digitsFound = [];
    let i = 0;
    while (i < line.length) {
        let matched = false;
        for (const [word, digit] of Object.entries(digitMap)) {
            if (line.startsWith(word, i)) {
                digitsFound.push(digit);
                i += word.length - 1; // Advance the index
                matched = true;
                break;
            }
        }
        i++; // Move to the next character if no match
    }

    return digitsFound;
}

function extractCalibrationValue(line) {
    const digits = extractDigits(line);
    if (digits.length > 0) {
        return parseInt(`${digits[0]}${digits[digits.length - 1]}`, 10);
    }
    return 0;
}

function calculateTotalCalibrationValue(filePath) {
    const lines = fs.readFileSync(filePath, 'utf-8').split('\n');
    return lines.reduce((sum, line) => sum + extractCalibrationValue(line), 0);
}

// Run the test
function test() {
    console.warn("Starting Tests")
    const testLines = [
        "two1nine",
        "eightwothree",
        "abcone2threexyz",
        "xtwone3four",
        "4nineeightseven2",
        "zoneight234",
        "7pqrstsixteen"
    ];
    const expectedResults = [29, 83, 13, 24, 42, 14, 76];

    testLines.forEach((line, index) => {
        const result = extractCalibrationValue(line);
        console.assert(result === expectedResults[index], `Error in line '${line}': Expected ${expectedResults[index]}, got ${result}`);
        console.log(`Line: '${line}', Calibration Value: ${result}`);
    });
    console.log("Finished Tests\n")
}

// Run the test function for verification
test();

// Calculate and print the total calibration value
const totalValue = calculateTotalCalibrationValue('../input.txt');
console.log(`Total Calibration Value from input.txt: ${totalValue}`);

File: Day01/cpp/solution.cpp
#include <iostream>
#include <fstream>
#include <sstream>
#include <string>
#include <vector>
#include <map>
#include <cassert>

std::vector<int> extractDigits(const std::string& line) {
    std::map<std::string, int> digitMap = {
        {"zero", 0}, {"one", 1}, {"two", 2}, {"three", 3}, {"four", 4},
        {"five", 5}, {"six", 6}, {"seven", 7}, {"eight", 8}, {"nine", 9}
    };
    for (int i = 0; i <= 9; ++i) {
        digitMap[std::to_string(i)] = i;
    }

    std::vector<int> digitsFound;
    size_t i = 0;
    while (i < line.length()) {
        bool matched = false;
        for (const auto& pair : digitMap) {
            if (line.substr(i, pair.first.length()) == pair.first) {
                digitsFound.push_back(pair.second);
                i += pair.first.length() - 1;
                matched = true;
                break;
            }
        }
        i++;
    }

    return digitsFound;
}

int extractCalibrationValue(const std::string& line) {
    std::vector<int> digits = extractDigits(line);
    if (!digits.empty()) {
        return digits.front() * 10 + digits.back();
    }
    return 0;
}

int calculateTotalCalibrationValue(const std::string& filePath) {
    std::ifstream file(filePath);
    if (!file.is_open()) {
        throw std::runtime_error("Unable to open file: " + filePath);
    }

    std::string line;
    int totalValue = 0;
    while (std::getline(file, line)) {
        totalValue += extractCalibrationValue(line);
    }

    return totalValue;
}

void test() {
    std::cout << "Starting Tests" << std::endl;
    std::vector<std::string> testLines = {
        "two1nine",
        "eightwothree",
        "abcone2threexyz",
        "xtwone3four",
        "4nineeightseven2",
        "zoneight234",
        "7pqrstsixteen"
    };
    std::vector<int> expectedResults = {29, 83, 13, 24, 42, 14, 76};

    for (size_t i = 0; i < testLines.size(); ++i) {
        int result = extractCalibrationValue(testLines[i]);
        assert(result == expectedResults[i]);
        std::cout << "Line: '" << testLines[i] << "', Calibration Value: " << result << std::endl;
    }
    std::cout << "Finished Tests" << std::endl << std::endl;
}

int main() {
    try {
        test();
        int totalValue = calculateTotalCalibrationValue("../input.txt");
        std::cout << "Total Calibration Value from input.txt: " << totalValue << std::endl;
    } catch (const std::exception& e) {
        std::cerr << "Error: " << e.what() << std::endl;
    }

    return 0;
}
